pkg/tagset/slice_builder.go

```diff
@@ -0,0 +1,186 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.Datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package tagset
+
+import (
+	"github.com/twmb/murmur3"
+)
+
+// SliceBuilder is used to build tagsets tag-by-tag, before "freezing" into a Tags instance. SliceBuilders are not threadsafe.
+//
+// A SliceBuilder goes through three stages in its lifecycle:
+//
+//     1. adding tags (begins on call to factory.NewSliceBuilder).
+//     2. frozen (begins on call to bldr.FreezeSlice).
+//     3. closed (begins on call to bldr.Close).
+//
+// The Add* methods may only be called in the "adding tags" stage. No methods
+// may be called in the "closed" stage.
+//
+// In general, the easiest way to ensure these stages are followed (and allow reviewers
+// to verify this) is to use a builder in a single method, as shown in the example.  Avoid
+// storing builders in structs.
+type SliceBuilder struct {
+	factory Factory
+
+	// Number of levels this builder supports
+	levels int
+
+	// Offsets into the tags/hashes arrays, such that level L is represented as
+	// tags[offsets[L]:offsets[L+1]]. This slice has length (levels+1).
+	offsets []int
```

Comment:
Instead of having `offsets` and a single `tags []string` why not having `tagsByLevel [][]string` (First dimension is the level index). I think it would make the implementation easier to understand with similar performance.

Reply:
The FreezeSlice operation may return tags from multiple levels, so if we collected tags in a two-level slice as you suggest, that data structure would later need to be "flattened" to a one-level slice to support FreezeSlice.  That'd be several additional allocations (one per level) and involve copying every element once, so I don't think it would be more efficient than the existing insertion-sort-based approach.

Reply:
Indeed, I did not notice there is no slice allocation in `FreezeSlice`